tools/autograd/gen_inplace_or_view_type.py

```diff
@@ -247,22 +285,139 @@ def emit_view_body(fn: NativeFunctionWithDifferentiabilityInfo, var: str) -> Tup
                      '/* is_fw_differentiable */ true, '
                      f'/* creation_meta */ {creation_meta});')
             rhs_value = f'std::move({var})'
         else:
             _, unpacked_bindings = unpack_args(f)
             call += emit_view_lambda(f, unpacked_bindings)
-            creation_meta = 'GradMode::is_enabled() ? CreationMeta::DEFAULT: CreationMeta::NO_GRAD_MODE'
-            rhs_value = (f'as_view(/* base */ {view_info}, /* output */ {var}, /* is_bw_differentiable */ true, '
+            creation_meta = ('at::GradMode::is_enabled() ? '
+                             'torch::autograd::CreationMeta::DEFAULT : '
+                             'torch::autograd::CreationMeta::NO_GRAD_MODE')
+            rhs_value = (f'torch::autograd::as_view(/* base */ {view_info}, /* output */ {var}, /* is_bw_differentiable */ true, '
                          '/* is_fw_differentiable */ true, '
                          f'/* view_func */ func, /* creation_meta */ {creation_meta})')
     else:
         # This could be supported but we don't need it at the moment, so keeping things simple.
         raise RuntimeError('Function that return multiple differentiable output '
                            'when at least one of them is view is not supported.')
     return call, rhs_value
+
+def emit_inplace_or_view_body(fn: NativeFunctionWithDifferentiabilityInfo) -> List[str]:
+    f = fn.func
+    inplace_view_body: List[str] = []
+
+    dispatcher_sig = DispatcherSignature.from_schema(f.func)
+    dispatcher_exprs = dispatcher_sig.exprs()
+
+    ret_and_arg_types = ', '.join([dispatcher_sig.returns_type()] + [a.type.cpp_type() for a in dispatcher_exprs])
+    # code-generated tracing kernels plumb and recompute dispatch keys directly through the kernel for performance.
+    # See Note [Plumbing Keys Through The Dispatcher] for details.
+    dispatch_key_set = 'ks & c10::DispatchKeySet(c10::DispatchKeySet::FULL_AFTER, c10::DispatchKey::InplaceOrView)'
+    redispatch_args = ', '.join([dispatch_key_set] + [a.expr for a in dispatcher_exprs])
+
+    # Note that this calls the slow, dispatching variants of manual_cpp_binding ops.
+    # We could probably work harder to ensure that the fast variants are called instead, but the perf benefit would be minimal.
+    sig_group = CppSignatureGroup.from_native_function(f, method=False, fallback_binding=f.manual_cpp_binding)
+    if sig_group.faithful_signature is not None:
+        api_name = sig_group.faithful_signature.name()
+    else:
+        api_name = sig_group.signature.name()
+    # inplace_view_body.append(THROW_IF_VARIABLETYPE_ON)
+    if modifies_arguments(f):  # inplace op
+        inplace_view_body.append(INPLACE_REDISPATCH.substitute(
+            api_name=api_name,
+            unpacked_args=redispatch_args,
+        ))
+        for r in cpp.return_names(f):
+            inplace_view_body.append(f'torch::autograd::increment_version({r});')
+    else:  # view op
```

Comment:
There are ops that are both view and inplace (`transpose_()` for example).

Reply:
I haven't the foggiest what `transpose_` is supposed to do. Does it restride the tensor? In that case, wouldn't you NOT version counter bump?

Reply:
You do want to version counter bump it. Because if an op saved the old version, it should fail as reading from the new Tensor (with updated strides) will return wrong result.

Note though that they are not properly handled today in the case of xla or cross-dtype view. But that will be covered in a separate issue: https://github.com/pytorch/pytorch/issues/49825#issuecomment-794466627
And this one will just aim at preserving the current behavior.

Reply:
Variable saving does a shallow copy:

```
Tensor VariableHooks::tensor_data(const Tensor& self) const {
  TORCH_CHECK(self.defined(), "cannot call tensor_data() on undefined tensor");
  auto self_impl_copy = self.unsafeGetTensorImpl()->shallow_copy_and_detach(
    /*version_counter=*/self.unsafeGetTensorImpl()->version_counter(),
    /*allow_tensor_metadata_change=*/self.unsafeGetTensorImpl()->allow_tensor_metadata_change());
  return at::Tensor(self_impl_copy);
} 
```

Therefore stride update will not be seen by the saved copy. https://www.youtube.com/watch?v=5-1ar64-Rr4

Reply:
So just to confirm, this means that any metadata-only op, inplace or not, can be treated as functional w.r.t. versioning, right? i.e. if backwards always reads these shallow copies

Reply:
Yes, because what version counter bump denotes (modulo detach) is a change to the *storage*; anything that isn't on storage (e.g., sizes, strides) doesn't count.